common/flockfile/src/lib.rs

```diff
@@ -0,0 +1,116 @@
+use log::{debug, warn};
+use nix::fcntl::{flock, FlockArg};
+use std::{
+    fs::{self, File, OpenOptions},
+    io,
+    os::unix::io::AsRawFd,
+    path::{Path, PathBuf},
+};
+
+#[derive(thiserror::Error, Debug)]
+pub enum FlockfileError {
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
+
+    #[error("Couldn't acquire file lock.")]
+    NixError(#[from] nix::Error),
+}
+
+/// flockfile creates a lockfile in the filesystem under `/run/lock` and then creates a filelock using system fcntl with flock.
+/// flockfile will automatically remove lockfile on application exit and the OS should cleanup the filelock afterwards.
+/// If application exits unexpectedly the filelock will be dropped, but the lockfile will not be removed unless handled in signal handler.
+#[derive(Debug)]
+pub struct Flockfile {
+    handle: Option<File>,
+    pub path: PathBuf,
+}
+
+impl Flockfile {
+    /// Create new lockfile in `/run/lock` with specific name:
+    ///
+    /// #Example
+    ///
+    /// let _lockfile = match flockfile::Flockfile::new_lock("app")).unwrap();
+    ///
+    pub fn new_lock(lock_name: impl AsRef<Path>) -> Result<Flockfile, FlockfileError> {
+        let path = Path::new("/run/lock").join(lock_name);
+
+        let file = OpenOptions::new()
+            .create(true)
+            .read(true)
+            .write(true)
+            .open(&path)?;
+
+        let () = flock(file.as_raw_fd(), FlockArg::LockExclusiveNonblock)?;
+
+        Ok(Flockfile {
+            handle: Some(file),
+            path,
+        })
+    }
+
+    /// Manually remove filelock and lockfile from the filesystem, this method doesn't have to be called explicitly,
+    /// however if access to the locked file is required this must be called.
+    pub fn unlock(mut self) -> Result<(), io::Error> {
+        self.handle.take().expect("handle dropped");
+        fs::remove_file(&self.path)?;
+        Ok(())
+    }
+}
+
+impl Drop for Flockfile {
+    fn drop(&mut self) {
+        if let Some(handle) = self.handle.take() {
+            drop(handle);
+
+            match fs::remove_file(&self.path) {
```

Comment:
I would add a comment + an integration test to make explicit that if drop is not called - the lock will be removed by the system even if the file is not removed and that this is not an issue if the file is created before an attempt to create the lock.

Reply:
The tests for this issues are in #232.